### PR TITLE
feat(cli): instrument cli with telemetry events

### PR DESCRIPTION
**Pull request by bjoerge**
### Description
Instruments the CLI with telemetry events

### What to review
- Does the setup make sense? Are the events on the format that we want them to be?

### Notes for release
- n/a will be included in general release notes for telemetry

### CODE REVIEW

**Comment by bjoerge** [comment]
Current dependencies on/for this PR:
* next
  * **PR #5321** <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/5321?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #5320** <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/5320?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #5322** <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/5322?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/sanity-io/sanity/5320?utm_source=stack-comment).

**Review by binoy14 — APPROVED** [review]

File: packages/@sanity/cli/src/__telemetry__/cli.telemetry.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  extraArguments: string[]
+  coreOptions: {
+    help: boolean
+    debug: boolean
+    version: boolean
+  }
+}
+
+export const CliCommand = defineTrace<CLITraceData>({
+  name: 'cliCommand',
```
Comment (joneidejohnsen) [inline packages/@sanity/cli/src/__telemetry__/cli.telemetry.ts:26]: @bjoerge This should still be "CLI Command Executed" right?

File: packages/sanity/src/_internal/cli/actions/build/build.telemetry.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import {defineTrace} from '@sanity/telemetry'
+
+export const BuildTrace = defineTrace<{outputSize: number}>({
+  name: 'Build Studio',
```
Comment (joneidejohnsen) [inline packages/sanity/src/_internal/cli/actions/build/build.telemetry.ts:4]: Should this be a separate event?

File: packages/@sanity/cli/src/__telemetry__/cli.telemetry.ts (hunk 1 above)
Comment (bjoerge) [reply packages/@sanity/cli/src/__telemetry__/cli.telemetry.ts:26]: Yes!

File: packages/@sanity/cli/src/cli.ts [hunk 3]
```diff
@@ -53,6 +70,17 @@ export async function runCli(cliRoot: string, {cliVersion}: {cliVersion: string}
     debug('No CLI config found')
   }
 
+  telemetry.updateUserProperties({
+    deviceId: await machineId(),
```
Comment (svirs) [inline packages/@sanity/cli/src/cli.ts:74]: This defaults to hashing the id 👍

File: packages/@sanity/cli/src/cli.ts [hunk 4]
```diff
@@ -87,13 +116,34 @@ export async function runCli(cliRoot: string, {cliVersion}: {cliVersion: string}
...
   const cliRunner = getCliRunner(commands)
-  cliRunner.runCommand(args.groupOrCommand, args, options).catch((err) => {
-    const error = typeof err.details === 'string' ? err.details : err
-    // eslint-disable-next-line no-console
-    console.error(`\n${error.stack ? neatStack(err) : error}`)
-    // eslint-disable-next-line no-process-exit
-    process.exit(1)
+  const cliCommandTrace = telemetry.trace(CliCommand)
+
+  cliCommandTrace.log({
```
Comment (svirs) [inline packages/@sanity/cli/src/cli.ts:121]: No `.start()` here

File: packages/@sanity/cli/src/actions/init-project/initProject.ts [hunk 5]
```diff
@@ -114,6 +126,21 @@ export default async function initSanity(
   let defaultConfig = cliFlags['dataset-default']
   let showDefaultConfigPrompt = !defaultConfig
 
+  trace.start()
```
Comment (svirs) [inline packages/@sanity/cli/src/actions/init-project/initProject.ts:129]: no trace.completed() for this one

File: packages/@sanity/cli/src/actions/login/login.ts [hunk 6]
```diff
@@ -43,25 +44,32 @@ export async function login(
...
   context: CliCommandContext,
 ): Promise<void> {
-  const {prompt, output, apiClient} = context
+  const {prompt, output, apiClient, telemetry} = context
   const {sso, experimental, open: openFlag, provider: specifiedProvider} = args.extOptions
   const previousToken = getCliToken()
   const hasExistingToken = Boolean(previousToken)
 
+  const trace = telemetry.trace(LoginTrace)
+  trace.start()
```
Comment (svirs) [inline packages/@sanity/cli/src/actions/login/login.ts:53]: corresponding trace.completed() missing

File: packages/@sanity/cli/src/actions/init-project/initProject.ts (hunk 5 above)
Comment (bjoerge) [reply packages/@sanity/cli/src/actions/init-project/initProject.ts:129]: Thanks! Added in 1f7574c

File: packages/@sanity/cli/src/cli.ts (hunk 4 above)
Comment (bjoerge) [reply packages/@sanity/cli/src/cli.ts:122]: Thanks! Added in 1f7574c

File: packages/@sanity/cli/src/actions/login/login.ts (hunk 6 above)
Comment (bjoerge) [reply packages/@sanity/cli/src/actions/login/login.ts:53]: Thanks! Added in 1f7574c

**Review by svirs — APPROVED** [review]
all tracing commands have a start/log(s)/complete calls with the right structure, looks good!

**Pull request merged**